### PR TITLE
Fix CUDA texture bugs and replace all instances of CUDA texture references with texture objects

### PR DESCRIPTION
**Pull request by cudawarped**
CUDA texture objects replaced CUDA texture references in the [2012](https://developer.nvidia.com/blog/cuda-pro-tip-kepler-texture-objects-improve-performance-and-flexibility/) (CUDA SDK 5.0 toolkit) and could be removed at any time from future SDK releases by Nvidia as they were depreciated in April 2021 (CUDA SDK 11.3).

This PR replaces all instances of texture references with texture objects using the existing updated `cv::cudev::Texture class`.

Additionaly it addresses the three bugs below related to incorrect texture usage:
1) `cv::cuda::resize` - currently resize uses texture references if the default stream or no stream is passed.  Under these circumstances global texture references are used when multiple threads call `cv::cuda::resize` which is unsafe.
2) `cv::cuda::demosaicing` - returns incorrect results when passed image ROI's.
3) `cv::cuda::HoughSegmentDetector` - passes the existing `cv::cudev::Texture` object whose destructor calls `cudaDestroyTextureObject()` to the CUDA kernel.  This can cause the underlying `cudaTextureObject_t` to be destroyed on return from the copy to the stub function before the kernel is launched as described in the [Nvidia docs](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#global-function-argument-processing).

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
buildworker:Custom=linux-1
docker_image:Custom=ubuntu-cuda:16.04
```

### CODE REVIEW

**Comment by alalek** [comment]
@asmorkalov How to run `Ubuntu2004-x64-CUDA` configuration on GitHub actions (GHA)?

File: modules/cudaimgproc/src/cuda/debayer.cu [hunk 1]
```diff
@@ -527,9 +527,15 @@ namespace cv { namespace cuda { namespace device
...
 
-        bindTexture(&sourceTex, src);
+        if (sourceOffset.x || sourceOffset.y) {
+            cv::cudev::TextureOff<uchar> texSrc(src, sourceOffset.y, sourceOffset.x);
+            MHCdemosaic < dst_t, cv::cudev::TextureOffPtr<uchar> > << <grid, block, 0, stream >> > ((PtrStepSz<dst_t>)dst, texSrc, firstRed);
+        }
+        else {
+            cv::cudev::Texture<uchar> texSrc(src);
+            MHCdemosaic < dst_t, cv::cudev::TexturePtr<uchar> > << <grid, block, 0, stream >> > ((PtrStepSz<dst_t>)dst, texSrc, firstRed);
+        }
```
Comment (asmorkalov) [inline modules/cudaimgproc/src/cuda/debayer.cu:537]: Issue with <> formatting.

File: modules/cudaimgproc/src/cuda/hough_segments.cu [hunk 2]
```diff
@@ -219,15 +220,18 @@ namespace cv { namespace cuda { namespace device
...
+            Point ofs;
+            mask.locateROI(wholeSize, ofs);
+            if (ofs.x || ofs.y) {
+                cv::cudev::TextureOff<uchar> texMask(wholeSize.height, wholeSize.width, mask.datastart, mask.step, ofs.y, ofs.x);
+                houghLinesProbabilistic<cv::cudev::TextureOffPtr<uchar>> << <grid, block, 0, stream >> > (texMask, accum, out, maxSize, rho, theta, lineGap, lineLength, mask.rows, mask.cols, counterPtr);
+            }
+            else {
+                cv::cudev::Texture<uchar> texMask(mask);
+                houghLinesProbabilistic<cv::cudev::TexturePtr<uchar>> << <grid, block, 0, stream >> > (texMask, accum, out, maxSize, rho, theta, lineGap, lineLength, mask.rows, mask.cols, counterPtr);
+            }
```
Comment (asmorkalov) [inline modules/cudaimgproc/src/cuda/hough_segments.cu:233]: Issue with <> formatting.

File: modules/cudaimgproc/test/test_color.cpp [hunk 3]
```diff
@@ -2419,7 +2420,7 @@ CUDA_TEST_P(Demosaicing, BayerBG2BGR_MHT)
     mosaic(img, src, cv::Point(1, 1));
 
     cv::cuda::GpuMat dst;
-    cv::cuda::demosaicing(loadMat(src), dst, cv::cuda::COLOR_BayerBG2BGR_MHT);
+    cv::cuda::demosaicing(loadMat(src,useRoi), dst, cv::cuda::COLOR_BayerBG2BGR_MHT);
```
Comment (asmorkalov) [inline modules/cudaimgproc/test/test_color.cpp:2423]: formatting issue.

File: modules/cudaimgproc/src/cuda/canny.cu [hunk 4]
```diff
@@ -151,63 +113,20 @@ namespace canny
...
-                L1 norm;
-                calcMagnitudeKernel<<<grid, block, 0, stream>>>(src, dx, dy, mag, norm);
-            }
-
-            cudaSafeCall( cudaGetLastError() );
-
-            if (stream == NULL)
-                cudaSafeCall( cudaDeviceSynchronize() );
+            L1 norm;
+            calcMagnitudeKernel << <grid, block, 0, stream >> > (texSrc, dx, dy, mag, norm);
```
[the remark this answers is not in the document]
Comment (cudawarped) [reply modules/cudaimgproc/src/cuda/canny.cu:125]: Noted, looks like Visual Studio copy paste madness!

**Review by asmorkalov — APPROVED** [review]

**Comment by cudawarped** [comment]
@asmorkalov I've just noticed I missed, [lut.cu](https://github.com/opencv/opencv_contrib/blob/4.x/modules/cudaarithm/src/cuda/lut.cu) and left the main texture reference routine in the main repo which is causing issues with CUDA 12.0 (https://github.com/opencv/opencv_contrib/issues/3390).  I'm addressing this now.

**Comment by asmorkalov** [comment]
Combination of changes in core and contrib leads to build error. CUDA: 10.2, Ubuntu 18.04.
```
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp: In member function ‘R cv::cudev::TexturePtr<T, R>::operator()(cv::cudev::TexturePtr<T, R>::index_type, cv::cudev::TexturePtr<T, R>::index_type) const’:
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:32:20: error: ‘tex2D’ was not declared in this scope
             return tex2D<R>(tex, x, y);
                    ^~~~~
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:32:20: note: suggested alternative: ‘tex’
             return tex2D<R>(tex, x, y);
                    ^~~~~
                    tex
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:32:27: error: expected primary-expression before ‘>’ token
             return tex2D<R>(tex, x, y);
                           ^
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp: In member function ‘R cv::cudev::TexturePtr<T, R>::operator()(cv::cudev::TexturePtr<T, R>::index_type) const’:
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:35:20: error: ‘tex1Dfetch’ was not declared in this scope
             return tex1Dfetch<R>(tex, x);
                    ^~~~~~~~~~
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:35:32: error: expected primary-expression before ‘>’ token
             return tex1Dfetch<R>(tex, x);
                                ^
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp: In member function ‘R cv::cudev::TexturePtr<long unsigned int, R>::operator()(cv::cudev::TexturePtr<long unsigned int, R>::index_type, cv::cudev::TexturePtr<long unsigned int, R>::index_type) const’:
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:48:34: error: ‘tex2D’ was not declared in this scope
             const uint2 retVal = tex2D<uint2>(tex, x, y);
                                  ^~~~~
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:48:34: note: suggested alternative: ‘tex’
             const uint2 retVal = tex2D<uint2>(tex, x, y);
                                  ^~~~~
                                  tex
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:48:45: error: expected primary-expression before ‘>’ token
             const uint2 retVal = tex2D<uint2>(tex, x, y);
                                             ^
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp: In member function ‘R cv::cudev::TexturePtr<long unsigned int, R>::operator()(cv::cudev::TexturePtr<long unsigned int, R>::index_type) const’:
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:52:34: error: ‘tex1Dfetch’ was not declared in this scope
             const uint2 retVal = tex1Dfetch<uint2>(tex, x);
                                  ^~~~~~~~~~
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:52:50: error: expected primary-expression before ‘>’ token
             const uint2 retVal = tex1Dfetch<uint2>(tex, x);
                                                  ^
In file included from /mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudaimgproc/src/precomp.hpp:62:0,
                 from /mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudaimgproc/src/bilateral_filter.cpp:43:
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp: In member function ‘R cv::cudev::TextureOffPtr<T, R>::operator()(cv::cudev::TextureOffPtr<T, R>::index_type, cv::cudev::TextureOffPtr<T, R>::index_type) const’:
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:65:20: error: ‘tex2D’ was not declared in this scope
             return tex2D<R>(tex, x + xoff, y + yoff);
                    ^~~~~
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:65:20: note: suggested alternative: ‘tex’
             return tex2D<R>(tex, x + xoff, y + yoff);
                    ^~~~~
                    tex
/mnt/projects/Projects/OpenCV/opencv_contrib/modules/cudev/include/opencv2/cudev/ptr2d/texture.hpp:65:27: error: expected primary-expression before ‘>’ token
             return tex2D<R>(tex, x + xoff, y + yoff);
```

**Pull request merged**